cartographer/mapping/id.h

```diff
@@ -322,6 +323,45 @@ class MapById {
 
   bool empty() const { return begin() == end(); }
 
+  // Returns an iterator to the the first element in the container belonging to
+  // trajectory 'trajectory_id' whose time is not considered to go before
+  // 'time', or EndOfTrajectory(trajectory_id) if all keys are considered to go
+  // before 'time'.
+  ConstIterator lower_bound(int trajectory_id, const common::Time& time) {
+    if (SizeOfTrajectoryOrZero(trajectory_id) == 0) {
+      return EndOfTrajectory(trajectory_id);
+    }
+
+    const std::map<int, DataType>& trajectory =
+        trajectories_.at(trajectory_id).data_;
+    if (std::prev(trajectory.end())->second.time() < time) {
+      return EndOfTrajectory(trajectory_id);
+    } else if (!(trajectory.begin()->second.time() < time)) {
+      return BeginOfTrajectory(trajectory_id);
+    }
+
+    int left = trajectory.begin()->first;
+    int right = std::prev(trajectory.end())->first;
+    while (true) {
+      const int middle = left + (right - left) / 2;
+      auto lower_bound_middle = trajectory.lower_bound(middle);
+      if (!(lower_bound_middle->second.time() < time)) {
+        if (std::prev(lower_bound_middle)->second.time() < time) {
```

Comment:
Can we remove this extra case? That is, change `left` and `right` to mean either left or right but maybe also at the solution. Since the interesting case for binary search is the condition that excludes a potential candidate, then do
```
if (std::prev(lower_bound_middle)->second.time() < time) {
  left = lower_bound_middle->first;
} else {
  right = std::prev(lower_bound_middle)->first;
}
```
and once `left == right` we are done.

Reply:
As discussed offline this is of course not correct. `middle` equals `left` in the interesting case, so it should be:
```
if (lower_bound_middle->second.time() < time) {
  left = std::next(lower_bound_middle)->first;
} else {
  right = lower_bound_middle->first;
}
```